### ISSUE TITLE
Signature by certificates with critical KeyUsage extension does not verify

### ISSUE THREAD

**Issue by KonstantinShemyak** [KonstantinShemyak/python-javatools#15]
When a certificate used for signing JAR contains
```
        X509v3 extensions:
            X509v3 Basic Constraints: 
                CA:FALSE
            X509v3 Key Usage: critical
                Digital Signature
            X509v3 Extended Key Usage: 
                Code Signing
```
Signature, made with such certificate, does not verify.

The reason *probably* is that *when* `keyUsage` extension is present, *then* it must contain `smimesign` in order for `PKCS7_verify()` to succeed.

Questions:
* does it depend on `critical`?
* does `ExtendedKeyUsage` have any effect?

**Comment by KonstantinShemyak** [comment]
The reason is identified correctly: `python-javatools` use `OpenSSL` for signature verification, namely `PKCS7_verify()`. The latter requires "smimesign purpose" ("purpose" here is an OpenSSL own term), which translates in certain `KeyUsage` extension combination. See http://securitypad.blogspot.fi/2017/01/openssl-pkcs7-verification-unsupported.html for details.

This could be overcome with setting custom verification callback for `PKCS7_verify()`. This is possible in `M2Crypto` since commit https://gitlab.com/m2crypto/m2crypto/commit/92f1120e6e45845ecd5c951c078a72eaa9c8ed75.

**Comment by KonstantinShemyak** [comment]
The problem is - **what** verification callback would be appropriate? Where is the specification, which X.503v3 extension combinations are "acceptable" for the JAR signature to be considered as "verified"?

**Comment by KonstantinShemyak** [comment]
`jarsigner` shows the following behavior:
- it also rejects signatures made with a certificate containing `KeyUsage`, but not `digitalSignature`:
```
$ jarsigner -verify -J-Djava.security.debug=jar -keystore no-dig-sig-bit-store.pfx -storepass 123456 data.jar 
TEST                                                                                                                          
jar: beginEntry META-INF/MANIFEST.MF
jar: done with meta!
jar: nothing to verify!
jar: beginEntry META-INF/MANIFEST.MF
jar: done with meta!
jar: nothing to verify!
jar: beginEntry META-INF/MANIFEST.MF
jar: done with meta!
jar: nothing to verify!
jar: beginEntry META-INF/MANIFEST.MF
jar: beginEntry META-INF/TEST.SF
jar: processEntry: processing block
jar: beginEntry META-INF/TEST.RSA
jar: processEntry: processing block
jar: processEntry caught: java.security.SignatureException: Key usage restricted: cannot be used for digital signatures
jar: done with meta!
jar: nothing to verify!

WARNING: Signature is either not parsable or not verifiable, and the jar will be treated as unsigned. For more information, re
-run jarsigner with debug enabled (-J-Djava.security.debug=jar).
```
- it accepts signatures made with a certificate containing `ExtendedKeyUsage`, but no `emailProtection` OID:

```
$ jarsigner -verify -J-Djava.security.debug=jar -keystore no-email-protection-store.pfx -storepass 1[114/4103]
.jar TEST                                                                                                                     
[...]
jar: beginEntry META-INF/TEST.RSA
jar: processEntry: processing block
[...]
jar verified.

Warning: 
This jar contains entries whose signer certificate's ExtendedKeyUsage extension doesn't allow code signing.
```
(Note the warning about missing `codeSigning` OID.)

The proposal for `python-javatools` is to mimic this behavior

**Comment by KonstantinShemyak** [comment]
Closed by  2542cc1.
